File: tools/utils/short-uuid.go
// License: GPLv3 Copyright: 2022, Kovid Goyal, <kovid at kovidgoyal.net>

package utils

import (
	"crypto/rand"
	"math"
	"math/big"

	"github.com/ALTree/bigfloat"
	"github.com/google/uuid"
)

const (
	ESCAPE_CODE_SAFE_ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~ "
	HUMAN_ALPHABET            = "23456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
)

func num_to_string(number *big.Int, alphabet []rune, alphabet_len *big.Int, pad_to_length int) string {
	var zero, digit big.Int
	if number.Sign() < 0 {
		*number = zero
	}
	capacity := 64
	if pad_to_length > capacity {
		capacity = pad_to_length
	}
	ans := make([]rune, 0, capacity)
	for number.Cmp(&zero) == 1 {
		number.DivMod(number, alphabet_len, &digit)
		ans = append(ans, alphabet[digit.Uint64()])
	}
	al := len(ans)
	if pad_to_length > -1 && al < pad_to_length {
		ans = ans[:pad_to_length]
		for i := al; i < pad_to_length; i++ {
			ans[i] = alphabet[0]
		}
	}
	return string(ans)
}

func get_padding_length(alphabet_len int) int {
	bi := big.NewInt(2)
	bi = bi.Exp(bi, big.NewInt(128), nil)
	bb := new(big.Float).SetPrec(256)
	bb.SetInt(bi)
	log_al := bigfloat.Log(big.NewFloat(float64(alphabet_len)).SetPrec(256))
	log_b := bigfloat.Log(bb)
	res := new(big.Float).SetPrec(256)
	res = res.Quo(log_b, log_al)
	val, _ := res.Float64()
	return int(math.Ceil(val))
}

type ShortUUID struct {
	alphabet      []rune
	alphabet_len  big.Int
	pad_to_length int
}

func CreateShortUUID(alphabet string) *ShortUUID {
	if alphabet == "" {
		alphabet = HUMAN_ALPHABET
	}
	var ans = ShortUUID{
		alphabet: []rune(alphabet),
	}
	ans.pad_to_length = get_padding_length(len(ans.alphabet))
	ans.alphabet_len.SetUint64(uint64(len(ans.alphabet)))
	return &ans
}

func (self *ShortUUID) Random(num_bits int64) (string, error) {
	max := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(num_bits), nil)
	bi, err := rand.Int(rand.Reader, max)
	if err != nil {
		return "", err
	}
	return num_to_string(bi, self.alphabet, &self.alphabet_len, self.pad_to_length), nil
}

func (self *ShortUUID) Uuid4() (string, error) {
	b, err := uuid.NewRandom()
	if err != nil {
		return "", err
	}
	bb, err := b.MarshalBinary()
	if err != nil {
		return "", err
	}
	var bi big.Int
	bi.SetBytes(bb)
	return num_to_string(&bi, self.alphabet, &self.alphabet_len, self.pad_to_length), nil
}

var HumanUUID *ShortUUID

func HumanUUID4() (string, error) {
	if HumanUUID == nil {
		HumanUUID = CreateShortUUID(HUMAN_ALPHABET)
	}
	return HumanUUID.Uuid4()
}

func HumanRandomId(num_bits int64) (string, error) {
	if HumanUUID == nil {
		HumanUUID = CreateShortUUID(HUMAN_ALPHABET)
	}
	return HumanUUID.Random(num_bits)
}

File: kitty/short_uuid.py
#!/usr/bin/env python
# License: GPLv3 Copyright: 2021, Kovid Goyal <kovid at kovidgoyal.net>

import math
import string
import uuid as _uuid
from collections.abc import Sequence


def num_to_string(number: int, alphabet: Sequence[str], alphabet_len: int, pad_to_length: int | None = None) -> str:
    ans = []
    number = max(0, number)
    while number:
        number, digit = divmod(number, alphabet_len)
        ans.append(alphabet[digit])
    if pad_to_length is not None and pad_to_length > len(ans):
        ans.append(alphabet[0] * (pad_to_length - len(ans)))
    return ''.join(ans)


def string_to_num(string: str, alphabet_map: dict[str, int], alphabet_len: int) -> int:
    ans = 0
    for char in reversed(string):
        ans = ans * alphabet_len + alphabet_map[char]
    return ans


escape_code_safe_alphabet = string.ascii_letters + string.digits + string.punctuation + ' '
human_alphabet = (string.digits + string.ascii_letters)[2:]


class ShortUUID:

    def __init__(self, alphabet: str = human_alphabet):
        self.alphabet = tuple(sorted(alphabet))
        self.alphabet_len = len(self.alphabet)
        self.alphabet_map = {c: i for i, c in enumerate(self.alphabet)}
        self.uuid_pad_len = int(math.ceil(math.log(1 << 128, self.alphabet_len)))

    def uuid4(self, pad_to_length: int | None = None) -> str:
        if pad_to_length is None:
            pad_to_length = self.uuid_pad_len
        return num_to_string(_uuid.uuid4().int, self.alphabet, self.alphabet_len, pad_to_length)

    def uuid5(self, namespace: _uuid.UUID, name: str, pad_to_length: int | None = None) -> str:
        if pad_to_length is None:
            pad_to_length = self.uuid_pad_len
        return num_to_string(_uuid.uuid5(namespace, name).int, self.alphabet, self.alphabet_len, pad_to_length)

    def decode(self, encoded: str) -> _uuid.UUID:
        return _uuid.UUID(int=string_to_num(encoded, self.alphabet_map, self.alphabet_len))


_global_instance = ShortUUID()
uuid4 = _global_instance.uuid4
uuid5 = _global_instance.uuid5
decode = _global_instance.decode
_escape_code_instance: ShortUUID | None = None


def uuid4_for_escape_code() -> str:
    global _escape_code_instance
    if _escape_code_instance is None:
        _escape_code_instance = ShortUUID(escape_code_safe_alphabet)
    return _escape_code_instance.uuid4()
